datalad_revolution/metadata/extractors/core.py

```diff
@@ -0,0 +1,302 @@
+            # TODO run `git log` to find earliest and latest commit to determine
+            # 'dateModified' and 'dateCreated'
+            # TODO determine per file 'contributor' from git log
+        }
+        return info
+
+    def get_state(self, dataset):
+        ds = dataset
+        return {
+            # increment when output format changes
+            'version': 1,
+            'unique_exclude': list(self._unique_exclude),
+            'remotes': ds.config.obtain(
+                'datalad.metadata.datalad-core.report-remotes',
+                True),
+            'contributors': ds.config.get(
+                'datalad.metadata.datalad-core.report-contributors',
+                True),
+            'modification-dates': ds.config.obtain(
+                'datalad.metadata.datalad-core.report-modification-dates',
+                True),
+        }
+
+
+def _get_file_key(rec):
+    # prefer the annex key, but fall back on the git shasum that is
+    # always around, identify the GITSHA as such in a similar manner
+    # to git-annex's style
+    return rec['key'] if 'key' in rec else 'SHA1-s{}--{}'.format(
+        rec['bytesize'] if rec['type'] != 'symlink' else 0,
+        rec['gitshasum'])
+
+
+def _get_commit_info(ds, status):
+    """Get info about all commits, up to (and incl. the refcommit)"""
+    #- get all the commit info with git log --pretty='%aN%x00%aI%x00%H'
+    #  - use all first-level paths other than .datalad and .git for the query
+    #- from this we can determine all modification timestamps, described refcommit
+    #- do a subsequent git log query for the determined refcommit to determine
+    #  a version by counting all commits since inception up to the refcommit
+    #  - we cannot use the first query, because it will be constrained by the
+    #    present paths that may not have existed previously at all
+
+    # determine the commit that we are describing
+    refcommit = get_refcommit(ds)
+    if refcommit is None or not len(status):
+        # this seems extreme, but without a single commit there is nothing
+        # we can have, or describe -> blow
+        # will turn into an error result upstairs
+        raise ValueError(
+            'No metadata-relevant repository content found. '
+            'Cannot determine reference commit for metadata ID'
+        )
+
+    # grab the history until the refcommit
+    stdout, stderr = ds.repo._git_custom_command(
+        None,
+        # name, email, timestamp, shasum
+        ['git', 'log', '--pretty=format:%aN%x00%aE%x00%aI%x00%H', refcommit],
+        expect_fail=True)
+    commits = [line.split('\0') for line in stdout.splitlines()]
+    # version, always anchored on the first commit (tags could move and
+    # make the integer commit count ambigous, and subtantially complicate
+    # version comparisons
+    version = '0-{}-g{}'.format(
+        len(commits),
+        # first seven chars of the shasum (like git-describe)
+        commits[0][3][:7],
```

Comment:
Probably doesn't matter here, but `git describe` doesn't hard code 7.  It respects `core.abbrev`, which with newer gits actually auto-scales based on the number of objects in the repository.  `git rev-parse --short COMMIT` (or `gitrepo.get_hexsha(..., short=True)` should give you the same abbreviation length `describe` would use.

Reply:
Awesome! Adjusted.